### PR TITLE
feat(core): bump tsp 0.51.0 and cadl-ranch 0.28.0

### LINKED ISSUE #4021: bump typespec `0.51.0` and cadl-ranch `0.28.0`

### PR DESCRIPTION
**Pull request by archerzz**
# Description

- bump typespec 0.51.0 and depending libraries
- bump cadl-ranch 0.28.0
- recursively load multiple-level client hierarchy
- regen tests/samples
- fix some tests

resolve #4021

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

**Comment by archerzz** [comment]
blocked by #4023

**Comment by archerzz** [comment]
blocked by https://github.com/Azure/autorest.csharp/pull/4019

**Comment by archerzz** [comment]
We're close to merging this PR. From the [latest preview PR](https://github.com/Azure/azure-sdk-for-net/pull/40784): 
- breaking change in `AnomalyDetector` is expected due to client hierarchy feature, and that SDK is deprecated
- breaking change in `healthinsights` should be fixed in upstream typespec fix: https://github.com/Azure/azure-rest-api-specs/pull/27081
- `devcenter` is fixed by restoring the logic of excluding return value for `delete` method

File: src/TypeSpec.Extension/Emitter.Csharp/src/lib/operation.ts [hunk 1]
```diff
@@ -364,14 +364,14 @@ export function loadOperation(
         }
 
         var bodyType = undefined;
-        if (op.verb !== "delete") {
+        if (
+            op.verb != "delete" &&
```
Comment (lirenhe) [inline src/TypeSpec.Extension/Emitter.Csharp/src/lib/operation.ts:368]: we should use  '!==' here?

Comment (archerzz) [reply src/TypeSpec.Extension/Emitter.Csharp/src/lib/operation.ts]: fixed

File: test/AutoRest.TestServer.Tests/ConvenienceMethodTests.cs [hunk 2]
```diff
@@ -238,10 +238,13 @@ public void NoConvenienceOptionalBodyScenario()
         [Test]
         public void SetKeepNonOverloadableProtocolSignature()
         {
-            var method = typeof(MixApiVersionClient).GetMethod("Delete");
+            var getPetsClientMethod = typeof(MixApiVersionClient).GetMethod("GetPetsClient");
```
Comment (lirenhe) [inline test/AutoRest.TestServer.Tests/ConvenienceMethodTests.cs:241]: Just curious, whey the emitter change would impact autorest tests?

Comment (pshao25) [reply test/AutoRest.TestServer.Tests/ConvenienceMethodTests.cs:241]: This is typespec code, not autorest, so expected. We should move these tests to some typespec places later.

**Review by lirenhe — APPROVED** [review]

**Pull request merged**